Compiler should hint to user to deref when doing reference assignment

https://is.gd/wm0S9G

``` rust
fn change_opt(opt: &mut Option<String>){
    opt = None
}
```

```
<anon>:2:11: 2:15 error: mismatched types:
 expected `&mut core::option::Option<collections::string::String>`,
    found `core::option::Option<_>`
(expected &-ptr,
    found enum `core::option::Option`) [E0308]
<anon>:2     opt = None
                   ^~~~
<anon>:2:11: 2:15 help: see the detailed explanation for E0308
error: aborting due to previous error
```

It'd be great if this had a hint:

```
 Did you mean `*opt = None`
```

Someone on IRC just ran into this.

The compiler now suggests:

```
error[E0308]: mismatched types
 --> src/main.rs:2:11
  |
2 |     opt = None
  |           ^^^^ expected mutable reference, found enum `std::option::Option`
  |
  = note: expected type `&mut std::option::Option<std::string::String>`
             found type `std::option::Option<_>`
  = help: try with `&mut None`
```

The suggestion is not smart enough to also suggest `*opt = None`.

Update:

```
error[E0308]: mismatched types
 --> src/main.rs:2:11
  |
2 |     opt = None
  |           ^^^^
  |           |
  |           expected mutable reference, found enum `std::option::Option`
  |           help: consider mutably borrowing here: `&mut None`
  |
  = note: expected type `&mut std::option::Option<std::string::String>`
             found type `std::option::Option<_>`
```

Compiler still not smart enough to suggest `*opt = None`.

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -306,11 +306,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     /// In addition of this check, it also checks between references mutability state. If the
     /// expected is mutable but the provided isn't, maybe we could just say "Hey, try with
     /// `&mut`!".
-    pub fn check_ref(&self,
-                 expr: &hir::Expr,
-                 checked_ty: Ty<'tcx>,
-                 expected: Ty<'tcx>)
-                 -> Option<(Span, &'static str, String)> {
+    pub fn check_ref(
+        &self,
+        expr: &hir::Expr,
+        checked_ty: Ty<'tcx>,
+        expected: Ty<'tcx>,
+    ) -> Option<(Span, &'static str, String)> {
         let cm = self.sess().source_map();
         let sp = expr.span;
         if !cm.span_to_filename(sp).is_real() {
@@ -397,6 +398,29 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                         } else {
                             String::new()
                         };
+                        if let Some(hir::Node::Expr(hir::Expr {
+                            node: hir::ExprKind::Assign(left_expr, _),
+                            ..
+                        })) = self.tcx.hir().find_by_hir_id(
+                            self.tcx.hir().get_parent_node_by_hir_id(expr.hir_id),
+                        ) {
+                            if mutability == hir::Mutability::MutMutable {
+                                // Found the following case:
+                                // fn foo(opt: &mut Option<String>){ opt = None }
+                                //                                   ---   ^^^^
+                                //                                   |     |
+                                //    consider dereferencing here: `*opt`  |
+                                // expected mutable reference, found enum `Option`
+                                if let Ok(src) = cm.span_to_snippet(left_expr.span) {
+                                    return Some((
+                                        left_expr.span,
+                                        "consider dereferencing here to assign to the mutable \
+                                         borrowed piece of memory",
+                                        format!("*{}", src),
+                                    ));
+                                }
+                            }
+                        }
                         return Some(match mutability {
                             hir::Mutability::MutMutable => (
                                 sp,

--- a/src/test/ui/suggestions/mut-ref-reassignment.rs
+++ b/src/test/ui/suggestions/mut-ref-reassignment.rs
@@ -1,5 +1,17 @@
-fn change_opt(opt: &mut Option<String>){
+fn suggestion(opt: &mut Option<String>) {
+    opt = None; //~ ERROR mismatched types
+}
+
+fn no_suggestion(opt: &mut Result<String, ()>) {
     opt = None //~ ERROR mismatched types
+}
+
+fn suggestion2(opt: &mut Option<String>) {
+    opt = Some(String::new())//~ ERROR mismatched types
+}
+
+fn no_suggestion2(opt: &mut Option<String>) {
+    opt = Some(42)//~ ERROR mismatched types
 }
 
 fn main() {}

--- a/src/test/ui/suggestions/mut-ref-reassignment.rs
+++ b/src/test/ui/suggestions/mut-ref-reassignment.rs
@@ -1,0 +1,5 @@
+fn change_opt(opt: &mut Option<String>){
+    opt = None //~ ERROR mismatched types
+}
+
+fn main() {}

--- a/src/test/ui/suggestions/mut-ref-reassignment.stderr
+++ b/src/test/ui/suggestions/mut-ref-reassignment.stderr
@@ -1,0 +1,16 @@
+error[E0308]: mismatched types
+  --> $DIR/mut-ref-reassignment.rs:2:11
+   |
+LL |     opt = None
+   |           ^^^^ expected mutable reference, found enum `std::option::Option`
+   |
+   = note: expected type `&mut std::option::Option<std::string::String>`
+              found type `std::option::Option<_>`
+help: consider dereferencing here to assign to the mutable borrowed piece of memory
+   |
+LL |     *opt = None
+   |     ^^^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

--- a/src/test/ui/suggestions/mut-ref-reassignment.stderr
+++ b/src/test/ui/suggestions/mut-ref-reassignment.stderr
@@ -1,16 +1,47 @@
 error[E0308]: mismatched types
   --> $DIR/mut-ref-reassignment.rs:2:11
    |
-LL |     opt = None
+LL |     opt = None;
    |           ^^^^ expected mutable reference, found enum `std::option::Option`
    |
    = note: expected type `&mut std::option::Option<std::string::String>`
               found type `std::option::Option<_>`
 help: consider dereferencing here to assign to the mutable borrowed piece of memory
    |
-LL |     *opt = None
+LL |     *opt = None;
    |     ^^^^
 
-error: aborting due to previous error
+error[E0308]: mismatched types
+  --> $DIR/mut-ref-reassignment.rs:6:11
+   |
+LL |     opt = None
+   |           ^^^^ expected mutable reference, found enum `std::option::Option`
+   |
+   = note: expected type `&mut std::result::Result<std::string::String, ()>`
+              found type `std::option::Option<_>`
+
+error[E0308]: mismatched types
+  --> $DIR/mut-ref-reassignment.rs:10:11
+   |
+LL |     opt = Some(String::new())
+   |           ^^^^^^^^^^^^^^^^^^^ expected mutable reference, found enum `std::option::Option`
+   |
+   = note: expected type `&mut std::option::Option<std::string::String>`
+              found type `std::option::Option<std::string::String>`
+help: consider dereferencing here to assign to the mutable borrowed piece of memory
+   |
+LL |     *opt = Some(String::new())
+   |     ^^^^
+
+error[E0308]: mismatched types
+  --> $DIR/mut-ref-reassignment.rs:14:11
+   |
+LL |     opt = Some(42)
+   |           ^^^^^^^^ expected mutable reference, found enum `std::option::Option`
+   |
+   = note: expected type `&mut std::option::Option<std::string::String>`
+              found type `std::option::Option<{integer}>`
+
+error: aborting due to 4 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

